### PR TITLE
[multitenancy-manager] Dedicated nodes / node groups for the Project

### PR DESCRIPTION
**Pull request by Dmitrykob**
## Description

We use the PodNodeSelector plugins to force node selection and PodTolerationRestriction to add tolerances to all pods in the project namespace.

screen1

screen2

## Why do we need it, and what problem does it solve?

Fix https://github.com/deckhouse/deckhouse/issues/4530

## What is the expected result?

All Pods of a project for which the value "dedicatedNodeName" is defined must run only on the Node with the corresponding taint key.
Taint must be applied manually beforehand.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: multitenancy-manager
type: fix
summary: All Pods of a project for which the value "dedicatedNodeName" is defined must run only on the Node with the corresponding taint key.
impact_level: default
```

### CODE REVIEW

File: ee/modules/160-multitenancy-manager/crds/projects.yaml [hunk 1]
```diff
@@ -54,9 +61,52 @@ spec:
...
+                    The name of the custom resource [ProjectType](cr.html#projecttype) to be used
+                    as a template for creating resources for the environment.
+                  type: string
+                template:
+                  additionalProperties:
+                    x-kubernetes-preserve-unknown-fields: true
+                  description:
+                    Values for [resource templates](cr.html#projecttype-v1alpha1-spec-resourcestemplate) from the environment template (CR [ProjectType](cr.html#projecttype) in helm values format that map to the OpenAPI specification (the [openAPI](cr.html#projecttype-v1alpha1-spec-openapi) parameter of the environment template).
+                  type: object
+                dedicatedNodeName:
```
Comment (yalosev) [inline ee/modules/160-multitenancy-manager/crds/projects.yaml:98]: we need some other contract here:
something like
```
dedicatedNodes:
  nodeSelector:
    matchLabels:
      foo: bar
  tolerations:
    - key: dedicated
      operator: Equal
      value: example
```

Comment (yalosev) [reply ee/modules/160-multitenancy-manager/crds/projects.yaml:98]:
```
            dedicatedNodes:
                type: object
                properties:
                    labelSelector:
                      description: A label selector is a label query over a set
                        of resources. The result of matchLabels and matchExpressions
                        are ANDed. An empty label selector matches all objects.
                        A null label selector matches no objects.
                      properties:
                        matchExpressions:
                          type: array
                          description: matchExpressions is a list of label selector
                            requirements. The requirements are ANDed.
                          items:
                            description: A label selector requirement is a selector
                              that contains values, a key, and an operator that
                              relates the key and values.
                            oneOf:
                              - properties:
                                  operator:
                                    enum: [Exists, DoesNotExist]
                                required: [key, operator]
                                not:
                                  required: [values]
                              - properties:
                                  operator:
                                    enum: [In, NotIn]
                                required: [key, operator, values]
                            properties:
                              key:
                                description: key is the label key that the selector
                                  applies to.
                                type: string
                              operator:
                                description: operator represents a key's relationship
                                  to a set of values. Valid operators are In, NotIn,
                                  Exists and DoesNotExist.
                                type: string
                              values:
                                description: values is an array of string values.
                                  If the operator is In or NotIn, the values array
                                  must be non-empty. If the operator is Exists or
                                  DoesNotExist, the values array must be empty.
                                  This array is replaced during a strategic merge
                                  patch.
                                type: array
                                items:
                                  type: string
                                  pattern: '[a-z0-9]([-a-z0-9]*[a-z0-9])?'
                                  minLength: 1
                                  maxLength: 63
                            type: object
                        matchLabels:
                          additionalProperties:
                            type: string
                          description: matchLabels is a map of {key,value} pairs.
                            A single {key,value} in the matchLabels map is equivalent
                            to an element of matchExpressions, whose key field is
                            "key", the operator is "In", and the values array contains
                            only "value". The requirements are ANDed.
                          type: object
                      type: object
                      x-kubernetes-map-type: atomic
```

File: ee/modules/160-multitenancy-manager/crds/doc-ru-projects.yaml [hunk 2]
```diff
@@ -19,6 +19,9 @@ spec:
                     Значения для [темплейтов ресурсов](cr.html#projecttype-v1alpha1-spec-resourcestemplate) из шаблона окружения (custom resource [ProjectType](cr.html#projecttype))
                     в формате helm values, которые соотносятся с
                     OpenAPI-спецификацией (поле [openAPI](cr.html#projecttype-v1alpha1-spec-openapi) шаблона окружения).
+                dedicatedNodeName:
+                  description: |
+                    Значение ключа (Taint) выделенной дял проекта ноды.
```
Comment (RomanenkoDenys) [inline ee/modules/160-multitenancy-manager/crds/doc-ru-projects.yaml:24]:
```suggestion
                    Значение ключа (Taint) выделенной для проекта ноды.
```

File: ee/modules/160-multitenancy-manager/webhooks/conversation/project [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+       .apiVersion = "deckhouse.io/v1alpha2"
+     else . end
+  )')
+  then
+    cat <<EOF >"$CONVERSION_RESPONSE_PATH"
+{"convertedObjects": $converted}
+EOF
+  else
+    cat <<EOF >"$CONVERSION_RESPONSE_PATH"
+{"failedMessage":"Conversion of nodegroups.deckhouse.io failed"}
```
Comment (RomanenkoDenys) [inline ee/modules/160-multitenancy-manager/webhooks/conversation/project:37]: I think nodegroups.deckhouse.io is wrong

File: ee/modules/160-multitenancy-manager/webhooks/conversation/project [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+       .apiVersion = "deckhouse.io/v1alpha1"
+     else . end
+  )')
+  then
+    cat <<EOF >"$CONVERSION_RESPONSE_PATH"
+{"convertedObjects": $converted}
+EOF
+  else
+    cat <<EOF >"$CONVERSION_RESPONSE_PATH"
+{"failedMessage":"Conversion of nodegroups.deckhouse.io failed"}
```
Comment (RomanenkoDenys) [inline ee/modules/160-multitenancy-manager/webhooks/conversation/project:54]: I think nodegroups.deckhouse.io is wrong

File: ee/modules/160-multitenancy-manager/webhooks/conversation/project
Comment (RomanenkoDenys) [inline ee/modules/160-multitenancy-manager/webhooks/conversation/project:1]: Why we need new api version for projects if nothing is changed in conversion webhook ?

**Review by RomanenkoDenys — CHANGES_REQUESTED** [review]
some questions

**Review by MariaMakeeva — APPROVED** [review]

**Review by MariaMakeeva — APPROVED** [review]

File: ee/modules/160-multitenancy-manager/webhooks/conversation/project (hunk 4 above)
Comment (Dmitrykob) [reply ee/modules/160-multitenancy-manager/webhooks/conversation/project]: conversation removed

File: ee/modules/160-multitenancy-manager/webhooks/conversation/project
Comment (Dmitrykob) [reply ee/modules/160-multitenancy-manager/webhooks/conversation/project:1]: conversation removed

File: ee/modules/160-multitenancy-manager/crds/doc-ru-projects.yaml (hunk 2 above)
Comment (Dmitrykob) [reply ee/modules/160-multitenancy-manager/crds/doc-ru-projects.yaml]: fixed

File: ee/modules/160-multitenancy-manager/templates/user-resources/user-resources-templates.yaml [hunk 5]
```diff
@@ -1,14 +1,16 @@
 {{- define "project.namespace" }}
 {{ $name := index . 0 }}
 {{ $metadata := index . 1 }}
+{{ $dedicatedNodeName := index . 2 }}
```
Comment (yalosev) [inline ee/modules/160-multitenancy-manager/templates/user-resources/user-resources-templates.yaml:4]: I think this variable must be renamed

Comment (Dmitrykob) [reply ee/modules/160-multitenancy-manager/templates/user-resources/user-resources-templates.yaml]: This viable is not used. So it must be deleted.

Comment (Dmitrykob) [reply ee/modules/160-multitenancy-manager/templates/user-resources/user-resources-templates.yaml]: Done

**Review by MariaMakeeva — APPROVED** [review]

File: ee/modules/160-multitenancy-manager/crds/projects.yaml (hunk 1 above)
Comment (Dmitrykob) [reply ee/modules/160-multitenancy-manager/crds/projects.yaml]: The contract is saved. 
The "dedicatedNodes" option has been moved to the template

File: ee/modules/160-multitenancy-manager/docs/USAGE_RU.md [hunk 6]
```diff
@@ -13,11 +13,21 @@ Deckhouse Kubernetes Platform включает следующий набор ш
   * выбор профиля безопасности;
   * настройка администраторов проекта.
 
-- `secure` — включает в себя все возможности шаблона `default` и дополнительные возможности:
+- `secure` — включает все возможности шаблона `default` и дополнительные функции:
   * настройку допустимых для проекта UID/GID;
   * правила аудита обращения к ядру линукс пользоваетелей проекта;
```
Comment (z9r5) [inline ee/modules/160-multitenancy-manager/docs/USAGE_RU.md:18]:
```suggestion
  * правила аудита обращения к ядру Linux-пользователей проекта;
```

File: ee/modules/160-multitenancy-manager/docs/USAGE_RU.md [hunk 7]
```diff
@@ -13,11 +13,21 @@ Deckhouse Kubernetes Platform включает следующий набор ш
...
   * сканирование запускаемых образов контейнеров на наличие CVE.
 
+- `secure-with-dedicated-nodes` — включает все возможности шаблона `secure` и дополнительные функции:
+  * определение селектора узла для всех подов в проекте: если под создан, селектор узла пода будет **заменен** автоматически селектором узла проекта;
+  * определение стандартных tolerations для всех подов в проекте: если под создан, стандартные tolerations будут **добавлена** к поду автоматически.
```
Comment (MariaMakeeva) [inline ee/modules/160-multitenancy-manager/docs/USAGE_RU.md:23]:
```suggestion
- `secure-with-dedicated-nodes` — включает все возможности шаблона `secure`, включая дополнительные функции:
  * определение селектора узла для всех подов в проекте: если под создан, селектор узла пода будет автоматически **заменён** на селектор узла проекта;
  * определение стандартных tolerations для всех подов в проекте:  если под создан, то для него стандартные значения tolerations добавляются автоматически.
```

Comment (z9r5) [inline ee/modules/160-multitenancy-manager/docs/USAGE_RU.md:23]:
```suggestion
  * определение стандартных tolerations для всех подов в проекте: если под создан, стандартные tolerations будут **добавлены** к поду автоматически.
```

**Review by MariaMakeeva — COMMENTED** [review]
I rewrote the text a bit

File: ee/modules/160-multitenancy-manager/docs/USAGE_RU.md (hunk 6 above)
Comment (MariaMakeeva) [reply ee/modules/160-multitenancy-manager/docs/USAGE_RU.md:18]: Обращение пользователей к ядру Linux или обращение именно Linux- пользователей? Мне кажется, двояко читается.

File: ee/modules/160-multitenancy-manager/docs/USAGE_RU.md [hunk 8]
```diff
@@ -13,10 +13,20 @@ Deckhouse Kubernetes Platform включает следующий набор ш
...
-- `secure` — включает в себя все возможности шаблона `default` и дополнительные возможности:
+- `secure` — включает все возможности шаблона `default` и дополнительные функции:
   * настройку допустимых для проекта UID/GID;
-  * правила аудита обращения к ядру линукс пользоваетелей проекта;
-  * сканирование запускаемых образов контейнеров на наличие CVE.
+  * правила аудита обращения Linux-пользователей проекта к ядру;
+  * сканирование запускаемых образов контейнеров на наличие известных уязвимостей (CVE).
+
+- `secure-with-dedicated-nodes` — расширяет возможности шаблона `secure`, а также дополнительные функции:
+  * определение селектора узла для всех подов в проекте: если под создан, селектор узла пода будет автоматически **заменён** на селектор узла проекта;
```
Comment (MariaMakeeva) [inline ee/modules/160-multitenancy-manager/docs/USAGE_RU.md:22]:
```suggestion
- `secure-with-dedicated-nodes` — включает все возможности шаблона `secure`, а также дополнительные функции:
  * определение селектора узла для всех подов в проекте: если под создан, селектор узла пода будет автоматически **заменён** на селектор узла проекта;
```

File: ee/modules/160-multitenancy-manager/docs/USAGE_RU.md [hunk 9]
```diff
@@ -13,10 +13,20 @@ Deckhouse Kubernetes Platform включает следующий набор ш
   * выбор профиля безопасности;
   * настройка администраторов проекта.
 
-- `secure` — включает в себя все возможности шаблона `default` и дополнительные возможности:
+- `secure` — включает все возможности шаблона `default` и дополнительные функции:
```
Comment (MariaMakeeva) [inline ee/modules/160-multitenancy-manager/docs/USAGE_RU.md:16]:
```suggestion
- `secure` — включает все возможности шаблона `default`, а также дополнительные функции:
```

**Review by MariaMakeeva — APPROVED** [review]
Fix text

**Pull request merged**